gadgets/trace_dns.bpf.c

```diff
@@ -0,0 +1,403 @@
+					       offsetof(struct tcphdr, dest));
+	} else if (event->src.proto == IPPROTO_UDP) {
+		event->src.port =
+			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
+					       offsetof(struct udphdr, source));
+		event->dst.port =
+			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
+					       offsetof(struct udphdr, dest));
+	}
+
+	event->qr = flags.qr;
+
+	if (flags.qr == 1) {
+		// Response code set only for replies.
+		event->rcode = flags.rcode;
+	}
+
+	bpf_skb_load_bytes(skb, DNS_OFF + sizeof(struct dnshdr), event->name,
+			   name_len);
+
+	event->pkt_type = skb->pkt_type;
+
+	// Read QTYPE right after the QNAME (name_len + the zero length octet)
+	// https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.2
+	event->qtype =
+		load_half(skb, DNS_OFF + sizeof(struct dnshdr) + name_len + 1);
+
+	// Enrich event with process metadata
+	struct sockets_value *skb_val = gadget_socket_lookup(skb);
+	if (skb_val != NULL) {
+		event->mntns_id = skb_val->mntns;
+		event->pid = skb_val->pid_tgid >> 32;
+		event->tid = (__u32)skb_val->pid_tgid;
+		__builtin_memcpy(&event->task, skb_val->task,
+				 sizeof(event->task));
+		event->uid = (__u32)skb_val->uid_gid;
+		event->gid = (__u32)(skb_val->uid_gid >> 32);
+	}
+
+	event->ancount = ancount;
+
+	// DNS answers start immediately after qname (name_len octets)
+	// + the zero length octet + qtype (2 octets) + qclass (2 octets).
+	int anoffset = DNS_OFF + sizeof(struct dnshdr) + name_len + 5;
+	int anaddrcount = load_addresses(skb, ancount, anoffset, event);
+	event->anaddrcount = anaddrcount;
+
+	// Calculate latency:
+	//
+	// Track the latency from when a query is sent from a container
+	// to when a response to the query is received by that same container.
+	//
+	// * On DNS query sent from a container namespace (qr == DNS_QR_QUERY and pkt_type == OUTGOING),
+	//   store the query timestamp in a map.
+	//
+	// * On DNS response received in the same container namespace (qr == DNS_QR_RESP and pkt_type == HOST)
+	//   retrieve/delete the query timestamp and set the latency field on the event.
+	//
+	// A garbage collection thread running in userspace periodically scans for keys with old timestamps
+	// to free space occupied by queries that never receive a response.
+	//
+	// Skip this if skb_val == NULL (gadget_socket_lookup did not set pid_tgid we use in the query key)
+	// or if event->timestamp == 0 (kernels before 5.8 don't support bpf_ktime_get_boot_ns, and the patched
+	// version IG injects always returns zero).
+	if (skb_val != NULL && event->timestamp > 0) {
+		struct query_key_t query_key = {
+			.pid_tgid = skb_val->pid_tgid,
+			.id = event->id,
+		};
+		if (event->qr == DNS_QR_QUERY &&
+		    event->pkt_type == PACKET_OUTGOING) {
+			bpf_map_update_elem(&query_map, &query_key,
+					    &event->timestamp, BPF_NOEXIST);
+		} else if (event->qr == DNS_QR_RESP &&
+			   event->pkt_type == PACKET_HOST) {
+			__u64 *query_ts =
+				bpf_map_lookup_elem(&query_map, &query_key);
```

Comment:
We need to find a way to clean this up

Reply:
I suggest to use bpf timers from Linux 5.15:
https://github.com/torvalds/linux/commit/b00628b1c7d595ae5b544e059c27b1f5828314b4

That would change the [requirements](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/requirements.md) for the dns gadget from Linux 5.4 to Linux 5.15. But the builtin dns gadget will still be here for a while, until users switch to Linux >= 5.15.